arwen/context/asyncCall.go

```diff
@@ -0,0 +1,286 @@
+		// if intraShard {
+		//   vmOutputWithError := createVMOutputInCaseOfBreakpointError(err)
+		//   return mergeTwoVMOutputs(intermediaryVMOutput, vmOutputWithError)
+		// }
+		return intermediaryVMOutput, nil
+	}
+
+	// Start calling the destination SC, synchronously.
+	destinationCallInput, err := host.createDestinationContractCallInput()
+	if err != nil {
+		vmOutputWithError := createVMOutputInCaseOfBreakpointError(err)
+		return mergeTwoVMOutputs(intermediaryVMOutput, vmOutputWithError), nil
+	}
+
+	destinationVMOutput, err := host.executeOnNewContextAndGetVMOutput(destinationCallInput)
+	// TODO handle this error properly
+	// TODO pass error to the SC callback (append as argument)
+	// TODO consume remaining gas
+	if err != nil {
+		vmOutputWithError := createVMOutputInCaseOfBreakpointError(err)
+		return mergeTwoVMOutputs(intermediaryVMOutput, vmOutputWithError), nil
+	}
+
+	intermediaryVMOutput = mergeTwoVMOutputs(intermediaryVMOutput, destinationVMOutput)
+
+	callbackCallInput, err := host.createCallbackContractCallInput(destinationVMOutput)
+	// TODO handle this error properly
+	if err != nil {
+		vmOutputWithError := createVMOutputInCaseOfBreakpointError(err)
+		return mergeTwoVMOutputs(intermediaryVMOutput, vmOutputWithError), nil
+	}
+
+	callbackVMOutput, err := host.executeOnNewContextAndGetVMOutput(callbackCallInput)
+	// TODO handle this error properly
+	if err != nil {
+		vmOutputWithError := createVMOutputInCaseOfBreakpointError(err)
+		return mergeTwoVMOutputs(intermediaryVMOutput, vmOutputWithError), nil
+	}
+	finalVMOutput := mergeTwoVMOutputs(intermediaryVMOutput, callbackVMOutput)
+
+	return finalVMOutput, nil
+}
+
+func (host *vmContext) createDestinationContractCallInput() (*vmcommon.ContractCallInput, error) {
+	sender := host.GetSCAddress()
+	dest := host.asyncCallDest
+	valueBytes := host.asyncCallValueBytes
+	data := host.asyncCallData
+	gasLimit := host.asyncCallGasLimit
+
+	err := host.argParser.ParseData(string(data))
+	if err != nil {
+		return nil, err
+	}
+
+	function, err := host.argParser.GetFunction()
+	if err != nil {
+		return nil, err
+	}
+
+	arguments, err := host.argParser.GetArguments()
+	if err != nil {
+		return nil, err
+	}
+
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  sender,
+			Arguments:   arguments,
+			CallValue:   big.NewInt(0).SetBytes(valueBytes),
+			GasPrice:    0,
+			GasProvided: gasLimit,
+		},
+		RecipientAddr: dest,
+		Function:      function,
+	}
+
+	return contractCallInput, nil
+}
+
+func (host *vmContext) createCallbackContractCallInput(destinationVMOutput *vmcommon.VMOutput) (*vmcommon.ContractCallInput, error) {
+	arguments := destinationVMOutput.ReturnData
+	gasLimit := destinationVMOutput.GasRemaining
+	function := "callBack"
+
+	// Calculate what length would the Data field have, were it of the
+	// form "callback@arg1@arg4...
+	dataLength := len(function) + 1
```

Comment:
Can we extract to a function called `getInputGasUsage` or something like that? And also write a unit test for it?

Reply:
Extracted as separate function, but unit test is pending.